### DISCUSSION TITLE
Trigger "Auf Änderung" triggert auch ohne Änderung?

### DISCUSSION THREAD

**Discussion by mipri25** [OpenCCU/OpenCCU#2001] [🇩🇪 Fragen&Antworten]
Hallo,

folgende Konstellation: ein ELV Charly mit aktueller RaspberryMatic, auf dem eine Homematic-Heizgruppe angelegt ist, der ein HM-CC-RT-DN und ein Fensterkontakt HM-Sec-SCo angehören.

Ich möchte nun ein Programm erstellen, das *einmalig* beim Schließen des Fensters ausgeführt wird. Dazu habe ich als Bedingung festgelegt:

WENN Geräteauswahl [der Fensterkontakt] 'bei' 'geschlossen' 'bei Änderung auslösen'

Leider wird das Programm damit aber nicht nur beim Schließen des Fensters ausgelöst, sondern auch immer dann, wenn der Fensterkontakt ca. alle 60 Minuten seinen Status meldet.

Meines Erachtens ist das nicht korrekt, denn 'bei Änderung auslösen' sollte ja nur dann triggern, wenn sich an der zugehörigen Bedingungszeile (hier also 'Fensterkontakt ist geschlossen') etwas ändert, und nachdem das Fenster bereits geschlossen war und der Fensterkontakt nur meldet, daß es nach wie vor geschlossen ist, liegt doch keine Änderung vor.

Das beobachtete Verhalten würde hingegen zu 'bei Aktualisierung' triggern passen.

Es macht auch keinen Unterschied, ob ich den Fensterkontakt aus der Heizgruppe oder das Gerät direkt verwende.

Mache ich hier einen Denkfehler, oder funktioniert das nicht wie vorgesehen?

Vielen Dank schonmal für jeden Tip!

**Comment by Baxxy13** [comment]
Das Programm triggert bei jeder "Aktualisierung" des Wertes, danach wird erst geprüft ob eine Änderung stattfand. 
Das was du beschreibst deutet darauf hin daß es entweder noch mehr Bedingungen, Sonst, wenn... oder sogar ein Sonst gibt. 
Ein Screenshot des vollständigen Programmes schafft da Klarheit.

**Reply by mipri25** [reply]
Vielen Dank für Deine Antwort. Eine weitere Bedingung ist nicht im Spiel, aber Dein Hinweis auf die zweigeteilte Abarbeitung des Triggers erklärt bereits meine Beobachtung, daß das Programm bei jeder Statusmeldung des Fensterkontakts erneut als ausgeführt markiert wird. Auf die Idee, daß dabei trotz erfüllter Wenn-Bedingung (das Fenster ist geschlossen) das 'Dann' gar nicht ausgeführt wird, wie ich inzwischen nachgeprüft habe, wäre ich so schnell allerdings nicht gekommen.

Offenbar ist die Reihenfolge der Abarbeitung also:

- Triggern wie bei "Auf Aktualisierung"
- Programm als ausgeführt markieren ("Letzte Ausführung" auf aktuelle Uhrzeit setzen)
- Erst dann prüfen, ob tatsächlich eine Änderung vorliegt
          - Fall nein: Abbruch der Bearbeitung, aber Programm bleibt als ausgeführt markiert.
          - Falls ja: Weiter mit Prüfung der "Wenn"-Bedingungen

Weil die Prüfung auf Änderung aber Teil der Trigger-Bedingung und nicht Teil der Wenn-Bedingung ist, ist diese Reihenfolge meines Erachtens nicht korrekt. Der Zeitpunkt der letzten Programmausführung sollte doch angeben, wann die Trigger-Bedingung das letzte Mal (komplett) erfüllt war und das Programm auch tatsächlich ausgeführt wurde. In meinem Beispiel also, wann das Fenster zuletzt geöffnet oder geschlossen wurde, und nicht, wann der Fensterkontakt zuletzt seinen (unveränderten) Status gesendet hat.

Demzufolge wäre die richtige Reihenfolge:

- Triggern wie bei "Auf Aktualisierung"
- Prüfen, ob tatsächlich auch eine Änderung vorliegt
       - Falls nein: Abbruch, ohne das Programm als ausfgeführt zu markieren
       - Falls ja: Programm als ausgeführt markieren und weiter mit Prüfung der Wenn-Bedingung

Ansonsten können einige nur schwer nachzuvollziehende Merkwürdigkeiten entstehen, über die auch ich hier gestolpert bin, z.B.

- Ein Programm wird als ausgeführt markiert, obwohl die Trigger-Bedingung gar nicht erfüllt war.
- Obwohl die Wenn-Bedingung zum Zeitpunkt der "Letzten Ausführung" erfüllt war, wurde das 'Dann' nicht ausgeführt.
- Das Programm wird als ausgeführt markiert, aber es wurde weder das 'Dann' noch das 'Sonst' ausgeführt.

Gibt es Gründe, die gegen eine Änderung der Reihenfolge sprechen? Ansonsten würde ich in den nächsten Tagen mal einen Issue erstellen...

**Comment by Baxxy13** [comment]
Du hättest das mal im Homematic-Forum vortragen sollen, da sind mehr Leute unterwegs die Ahnung haben und auch ihre Meinung abgeben.

Richtig oder falsch ist auch Ansichtssache. 
Letztlich "war es schon immer so" und man lernt damit umzugehen. 
Besser gesagt man sollte lernen damit umzugehen denn ändern wird RaspberryMatic daran sicher nichts.

Wenn man sein System im Griff hat ist der Zeitstempel bei "Letzte Ausführung" auch relativ uninteressant. 
Wenn wir ihn jetzt einfach in "letzte Programmtriggerung" umbenennen wäre doch alles paletti, oder? ;-)

**Reply by MichaelN0815** [reply]
Ich finde die Idee gar nicht so doof. Ja, man lernt damit umzugehen. Aber wäre es nicht besser das Problem zu beseitigen?

Wir nutzen Protokoll Variablen oder Alchys Auslöser Skript um dieses issue zu umgehen.

Wie cool wäre es, wenn RaspberryMatic von sich aus nicht nur den Auslöse Zeitpunkt, sondern auch den Trigger anzeigen würde?

**Reply by mipri25** [reply]
Vielen Dank auch hier für Deine Antwort. Du hast insofern recht, als daß die Diskussion im Homematic-Forum vermutlich besser aufgehoben gewesen wäre. Da wir nun aber schonmal hier sind, noch ein paar Anmerkungen, denn ich finde ganz und gar nicht, daß mit der Umbenennung, die Du vorschlägst, alles in Ordnung wäre.

Grund: Die Überpüfung auf Änderung ist Teil der Triggerbedingung. Ohne tatsächlich erfolgte Änderung, ist die Triggerbedingung also nicht erfüllt. Und wenn die Triggerbedingung nicht erfüllt ist, darf mit dem Programm garnichts passieren.

Die jetzige Angabe der "Letzten Programmausführung" gibt lediglich den Zeitpunkt wider, wann die Triggerbedingung zuletzt geprüft wurde, ohne daß man weiß, wie das Ergebnis war. Das ist in etwa so unsinnig, als wenn ein Programm, das zu einem bestimmten Zeitpunkt ausgeführt werden soll, jede Minute als ausgeführt markiert würde, nur weil da geprüft wurde, ob die Ausführungs-Uhrzeit bereits erreicht ist. Und wenn jemand diese Angabe tatsächlich braucht, dann kann man ja "Auf Aktualisierung" triggern und selber auf Änderung prüfen.

Und ganz allgemein noch: die Tatsache, daß etwas schon immer so war, ist nie eine hinreichende Begründung dafür, es nicht ab sofort besser zu machen. Und aus eigener Erfahrung als Entwickler weiß ich, daß es sich *immer* lohnt, die Liste der Dinge, die man "halt wissen muß", weil in Wahrheit irgendwas nicht funktioniert wie vorgesehen, möglichst kurz zu halten, weil das dem Gesamteindruck des Ganzen ungemein zugute kommt. Selbst dann, wenn es sich um vermeintliche Kleinigkeiten handelt.

**Reply by Baxxy13** [reply]
> denn ich finde ganz und gar nicht, daß mit der Umbenennung, die Du vorschlägst, alles in Ordnung wäre

Ein Zwinker-Smiley bedeutet eigentlich das man das Geschriebene nicht ganz so ernst nehmen sollte.
Mit dieser "Namensänderung" wäre natürlich nicht alles in Ordnung. Es würde m.E. aber besser zum IST-Zustand passen. 

> Grund: Die Überpüfung auf Änderung ist Teil der Triggerbedingung.

Darüber wurde schon mehrfach gestritten und gefachsimpelt. 
Der Passus im WebUI-Handbuch lässt da (zumindest für mich) einigen Interpretationsspielraum.

> Die Trigger-Ereignisse sind komplett losgelöst von den logischen Abfrage-Bedingungen der
> Programme zu betrachten und sind nicht selbst Bestandteil der Bedingung. Dies bedeutet,
> dass z. B. die Bedingung „Wenn: Schaltaktor eingeschaltet, Auslösen auf Änderung“ das
> Programm sowohl beim Einschalten, als auch beim Ausschalten triggert, die Bedingung
> aber während der gesamtem Zeit, in der der Aktor eingeschaltet ist, erfüllt ist. Würde das
> Programm also nach dem Einschalten des Aktors also durch ein anderes Ereignis nochmals
> getriggert, wäre diese Bedingung also erfüllt (weil der Aktor eingeschaltet ist). Das Ereignis
> „Änderung“ ist selbst also nicht Teil der logischen Abfrage, sondern dient lediglich als ein Auslöser,
> um die Prüfung und Abarbeitung des gesamten Programms zu starten.

Mit meinen Worten:
Jedes "zucken" eines Wertes (echte Änderung oder Aktualisierung) triggert das Programm. 
Steht der Trigger-Wert "auf Aktualisierung" wird sofort mit der Bedingungsprüfung begonnen.
Steht der Trigger-Wert "auf Änderung" wird erst "geprüft" ob es eine Änderung gab und wenn ja beginnt die Bedingungsprüfung.

Man kann das Prozedere sehr gut nachvollziehen wenn man mal die Logik auf "alles loggen" stellt. Idealerweise auf einem "nackten System" ohne Aktoren/Sensoren/Funkmodul mit nur einem Testprogramm und zwei Test-SysVars. Die Analyse des Logfiles ist zwar etwas aufwändig aber aufschlußreich. 

> Die jetzige Angabe der "Letzten Programmausführung" gibt lediglich den Zeitpunkt wider, wann die Triggerbedingung zuletzt geprüft wurde, ohne daß man weiß, wie das Ergebnis war.

Wie sollte man das anders handhaben? Wenn du bspw. 10 DANN-Pfade und ein SONST hast bekommst du doch auch "nur" einen Zeitstempel und weißt trotzdem nicht wie das Ergebnis war. War es eins der DANN's oder doch das SONST?

> es nicht ab sofort besser zu machen

Und genau hier kommen wir zum Problem. Wer soll das machen? eQ-3? Jens? Es gibt nicht viele die das Konstrukt "CCU" vollends überblicken, noch weniger die die Fähigkeiten haben was zu ändern und quasi keinen der Lust hat eine Baustelle aufzumachen.
Hinzu kommt das einige Teile Closed-Source sind.

Sicherlich "könnte" man das besser/anders machen, aber sollte man das wirklich?
Die "Erfahrenen" haben die Logik intus, die gucken einmal über ein Programm und sehen quasi sofort was der Neuling da verbockt hat.
Eine Nachfrage nach Zeitstempeln kommt eher selten weil die "Erfahrenen" wissen das diese nur "Schall und Rauch" sind.
Außerdem gibt es etablierte Methoden um Programmausführungen und Triggerereignisse zu protokollieren.

Sorry das ich das so sagen muss, aber für mich stünde hier der Aufwand in keinem Verhältnis zum Nutzen.

**Reply by mipri25** [reply]
> Ein Zwinker-Smiley bedeutet eigentlich das man das Geschriebene nicht ganz so ernst nehmen sollte.

Da hast Du recht. Aber ich habe mich ehrlich gesagt ein bißchen über die Floskeln wie "das war schon immer so", "damit muß man leben", "das stört nur, wenn man sein System nicht im Griff hat" geärgert und es auch deshalb etwas ernster genommen als nötig.

> Der Passus im WebUI-Handbuch lässt da (zumindest für mich) einigen Interpretationsspielraum.

Der Passus, den Du zitierst, sagt für mich einfach nur, daß z.B. folgendes (natürlich unsinnige) Programm

WENN Geräteauswahl [ein Fensterkontakt] 'bei geschlossen' 'bei Änderung ausführen'
UND
Zeitsteuerung 'um 14:00 ausführen' 'zu Zeitpunkten auslösen'

um 14:00 gestartet wird und das 'Dann' ausführt, auch wenn es beim Zustand des Fensterkontakts zu diesem Zeitpunkt keine Änderung gab. Das Fenster muß einfach nur um 14:00 geschlossen sein.

> Mit meinen Worten:
Jedes "zucken" eines Wertes (echte Änderung oder Aktualisierung) triggert das Programm.
Steht der Trigger-Wert "auf Aktualisierung" wird sofort mit der Bedingungsprüfung begonnen.
Steht der Trigger-Wert "auf Änderung" wird erst "geprüft" ob es eine Änderung gab und wenn ja beginnt die Bedingungsprüfung.

Das sehe ich fast genauso. Für mich triggert das Programm aber erst, wenn die Trigger-Bedingung erfüllt ist. Bei "Auf Änderung" also erst, wenn tatsächlich eine Änderung festgestellt wurde. Dann beginnt die Programmausführung mit der Prüfung der Wenn-Bedingungen, und somit sollte auch erst hier der Ausführungszeitpunkt gesetzt bzw. aktualisiert werden.

Im Prinzip ist das aber egal. Auch bei Deiner Betrachtungsweise wäre alles in Ordnung, wenn die letzte Programmausführung erst und ausschließlich zu Beginn der Bedingungprüfung gesetzt würde.

>>Die jetzige Angabe der "Letzten Programmausführung" gibt lediglich den Zeitpunkt wider, wann die Triggerbedingung zuletzt geprüft wurde, ohne daß man weiß, wie das Ergebnis war.

> Wie sollte man das anders handhaben?

Indem man den Zeitpunkt der letzten Programmausführung erst setzt, nachdem(!) die Prüfung der Triggerbedingung positiv ausgefallen ist und es auch wirklich zur Programmausführung kommt, also zur Prüfung der Wenn-Bedingung und somit zur Ausführung des Dann oder des Sonst. So wie es jetzt ist, kann zum angegebenen Zeitpunkt der letzten Programmausführung ja auch einfach nichts passiert sein, weil schon die Triggerbedingung nicht erfüllt war.

Welchen Weg durch die Danns und Sonsts das Programm bei seiner Ausführung dann genommen hat, geht daraus natürlich nicht hervor, aber man weiß immerhin, daß es überhaupt eine Ausführung gab.

> Und genau hier kommen wir zum Problem. Wer soll das machen? eQ-3? Jens? Es gibt nicht viele die das Konstrukt "CCU" vollends überblicken, noch weniger die die Fähigkeiten haben was zu ändern und quasi keinen der Lust hat eine Baustelle aufzumachen.
Hinzu kommt das einige Teile Closed-Source sind.

Gut, das sind jetzt Begründungen, die nachvollziehbar sind. Das konnte ich aber nicht einschätzen, und deshalb habe ich zuerst hier nachgefragt, bevor ich einen Issue erstellt habe. Ich war davon ausgegangen, daß es einen (viel) größeren Kreis an Entwicklern gibt, und daß es für jemand, der mit dem Code vertraut ist, eher eine Kleinigkeit ist, das Setzen des Zeitpunkts der letzten Programmausführung von vor dem Prüfen auf Aktualiserung nach hinter diese Prüfung zu verschieben.

> Sorry das ich das so sagen muss, aber für mich stünde hier der Aufwand in keinem Verhältnis zum Nutzen.

Kein Grund für sorry. Wenn es sich nicht mit vertretbarem Aufwand ändern läßt, muß man das so akzeptieren und kann dann natürlich auch ohne weiteres damit leben. Aber ich wollte es zumindest versucht haben.

Ich danke Dir jedenfalls sehr für Deine Erklärungen und die Zeit, die Du Dir dafür genommen hast,
